File: nodes/random-number/random-number.process.py
import random
import math

async def process(inputs, settings, config, nodeConfig):
    """
    Process function for the Random Number Generator node.
    Generates random numbers with configurable range and distribution.
    """
    # Create a JavaScript-compatible random number generator if seed is provided
    if "seed" in inputs and inputs["seed"] is not None:
        seed_value = float(inputs["seed"])
        random_func = create_seeded_random(seed_value)
    else:
        # Use Python's built-in random for unseeded random numbers
        random_func = random.random
    
    distribution = inputs.get("distribution", "uniform")
    
    if distribution == "normal":
        # Implement Box-Muller transform for normal distribution
        mean = float(inputs.get("mean", 0))
        stddev = float(inputs.get("stddev", 1))
        
        # Box-Muller transform to generate normal distribution from uniform
        u1 = random_func()
        u2 = random_func()
        z0 = math.sqrt(-2.0 * math.log(u1)) * math.cos(2.0 * math.pi * u2)
        
        value = mean + stddev * z0
    
    elif distribution == "integer":
        min_val = int(inputs.get("min", 0))
        max_val = int(inputs.get("max", 100))
        
        # JavaScript's Math.random() is [0, 1), so we need to adjust
        value = min_val + math.floor(random_func() * (max_val - min_val + 1))
    
    else:  # uniform distribution
        min_val = float(inputs.get("min", 0))
        max_val = float(inputs.get("max", 1))
        
        value = min_val + random_func() * (max_val - min_val)
    
    return {"value": value}

def create_seeded_random(seed):
    """
    A simple seeded random number generator based on a linear congruential generator
    This exactly matches the JavaScript implementation
    """
    # Constants for a simple Linear Congruential Generator
    a = 1664525
    c = 1013904223
    m = 2**32
    
    current_seed = seed
    
    def random_generator():
        nonlocal current_seed
        # Update the seed
        current_seed = (a * current_seed + c) % m
        # Return a number between 0 and 1
        return current_seed / m
    
    return random_generator 

File: nodes/random-number/random-number.process.js
/**
 * A simple seeded random number generator based on a linear congruential generator
 * @param {number} seed - The seed value
 * @returns {function} A function that returns a random number between 0 and 1
 */
function createSeededRandom(seed) {
  // Constants for a simple Linear Congruential Generator
  const a = 1664525;
  const c = 1013904223;
  const m = Math.pow(2, 32);
  
  let currentSeed = seed;
  
  return function() {
    // Update the seed
    currentSeed = (a * currentSeed + c) % m;
    // Return a number between 0 and 1
    return currentSeed / m;
  };
}

export default async ({inputs, settings, config}) => {

  // Determine which random function to use
  let random = Math.random;
  
  // Use seeded random if a seed is provided
  if (inputs.seed !== undefined) {
    const seed = Number(inputs.seed);
    random = createSeededRandom(seed);
  }

  let value;
  const distribution = inputs.distribution || 'uniform';
  
  switch (distribution) {
    case 'normal': {
      // Box-Muller transform for normal distribution
      const mean = Number(inputs.mean) || 0;
      const stddev = Number(inputs.stddev) || 1;
      
      const u1 = random();
      const u2 = random();
      
      const z0 = Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2);
      value = z0 * stddev + mean;
      break;
    }
    
    case 'integer': {
      const min = Math.ceil(Number(inputs.min) || 0);
      const max = Math.floor(Number(inputs.max) || 100);
      value = Math.floor(random() * (max - min + 1)) + min;
      break;
    }
    
    case 'uniform':
    default: {
      const min = Number(inputs.min) || 0;
      const max = Number(inputs.max) || 1;
      value = random() * (max - min) + min;
      break;
    }
  }

  return { value };
}; 